test/aleph/resource_leak_detector.clj

```diff
@@ -0,0 +1,210 @@
+  (swap! current-leaks conj {:resource-type resource-type
+                             :records records}))
+
+;; NOTE: Since we require level PARANOID, this should never be called in practice.
+(defn -reportUntracedLeak [_this resource-type]
+  (swap! current-leaks conj {:resource-type resource-type
+                             :records "[untraced]"}))
+
+(defn log-leaks! [leaks]
+  (doseq [{:keys [resource-type records]} leaks]
+    ;; Log message cribbed from io.netty.util.ResourceLeakDetector's (protected) reportTracedLeak method
+    (log/error (str "LEAK: " resource-type ".release() was not called before it's garbage-collected.")
+               (str "See https://netty.io/wiki/reference-counted-objects.html for more information." records))))
+
+(defmacro with-expected-leaks
+  "Runs `body` and expects it to produce exactly `expected-leak-count` leaks. Intended for use in tests
+  which intentionally leak resources.
+
+  Requires the leak detector to be `enabled?`."
+  [expected-leak-count & body]
+  `(with-leak-collection
+     (fn [] ~@body)
+     ;; NOTE: Using a raw symbol here instead of a gensym to get nicer test failures.
+     (fn [~'leaks]
+       (when-not (test/is (= ~expected-leak-count (count ~'leaks)) "Unexpected leak count! See log output for details.")
+         (log-leaks! ~'leaks)))))
+
+(defn- report-test-leaks! [leaks]
+  (when (seq leaks)
+    (log-leaks! leaks)
+    ;; We include the assertion here within the `when` form so that we don't add a mystery assertion
+    ;; to every passing test (which is the common case).
+    (test/is (zero? (count leaks))
+             "Leak detected! See log output for details.")))
+
+(defn- instrument-test-fn [tf]
+  (if (::instrumented? tf)
+    tf
+    (with-meta
+      (fn []
+        (with-leak-collection tf report-test-leaks!))
+      {::instrumented? true})))
+
+(defn instrument-tests!
+  "If `enabled?`, instruments all tests in the current namespace with leak detection by wrapping them
+  in `with-leak-collection`. If leaks are detected, a corresponding (failing) assertion is injected
+  into the test and the leak reports are logged at level `error`.
+
+  Usually placed at the end of a test namespace.
+
+  Note that this is intentionally not implemented as a fixture since there is no clean way to make a
+  test fail from within a fixture: Neither a failing assertion nor throwing an exception will
+  preserve which particular test caused it. See
+  e.g. https://github.com/technomancy/leiningen/issues/2694 for an example of this."
+  []
+  (when (enabled?)
+    (->> (ns-interns *ns*)
+         vals
+         (filter (comp :test meta))
+         (run! (fn [tv]
+                 (alter-meta! tv update :test instrument-test-fn))))))
+
+(if (enabled?)
+  (log/info "enabled.")
+  (log/info "disabled. This means resource leaks will be reported less accurately."
+            "Pass `-Dio.netty.customResourceLeakDetector=aleph.resource_leak_detector` to enable."))
```

Comment:
Minor: add something useful before "enabled/disabled" in case the log settings don't display the namespace.